### PR TITLE
Reapply "[Clang][CWG1815] Support lifetime extension of temporary created by aggregate initialization using a default member initializer"

### PR DESCRIPTION
**Pull request by yronglin**
This PR reapply https://github.com/llvm/llvm-project/pull/87933

### CODE REVIEW

File: clang/lib/Sema/SemaExpr.cpp [hunk 1]
```diff
@@ -5695,19 +5694,35 @@ ExprResult Sema::BuildCXXDefaultInitExpr(SourceLocation Loc, FieldDecl *Field) {
...
         NestedDefaultChecking;
-
+    // Pass down lifetime extending flag, and collect temporaries in
+    // CreateMaterializeTemporaryExpr when we rewrite the call argument.
+    keepInLifetimeExtendingContext();
     EnsureImmediateInvocationInDefaultArgs Immediate(*this);
     ExprResult Res;
+
+    // Rebuild CXXDefaultInitExpr might cause diagnostics.
+    SFINAETrap Trap(*this);
```
Comment (yronglin) [inline clang/lib/Sema/SemaExpr.cpp:5720]: Should we suppress diagnostic messages? Perhaps the initializer with errors should not be rebuilt.

Comment (yronglin) [reply clang/lib/Sema/SemaExpr.cpp:5720]: Is there a way to detect if there is an error in the initializer? For example, access errors.

File: clang/test/SemaCXX/cxx11-default-member-initializers.cpp [hunk 2]
```diff
@@ -27,6 +27,80 @@ class MemInit {
...
+  initializer_list() : __begin_(nullptr), __size_(0) {}
+
+  size_t size() const { return __size_; }
+  const _E *begin() const { return __begin_; }
+  const _E *end() const { return __begin_ + __size_; }
+};
+} // namespace std
+
+#if __cplusplus >= 201703L
+namespace test_rebuild {
```
Comment (yronglin) [inline clang/test/SemaCXX/cxx11-default-member-initializers.cpp:58]: Add more test that from https://github.com/llvm/llvm-project/pull/87933 's comments.

File: clang/lib/Sema/TreeTransform.h [hunk 3]
```diff
@@ -14113,12 +14120,9 @@ TreeTransform<Derived>::TransformCXXTemporaryObjectExpr(
     return SemaRef.MaybeBindToTemporary(E);
   }
 
-  // FIXME: We should just pass E->isListInitialization(), but we're not
-  // prepared to handle list-initialization without a child InitListExpr.
   SourceLocation LParenLoc = T->getTypeLoc().getEndLoc();
   return getDerived().RebuildCXXTemporaryObjectExpr(
-      T, LParenLoc, Args, E->getEndLoc(),
-      /*ListInitialization=*/LParenLoc.isInvalid());
+      T, LParenLoc, Args, E->getEndLoc(), E->isListInitialization());
```
Comment (yronglin) [inline clang/lib/Sema/TreeTransform.h:14125]: The issue in https://github.com/llvm/llvm-project/pull/87933 caused by `/*ListInitialization=*/LParenLoc.isInvalid()`.

File: clang/lib/Sema/TreeTransform.h [hunk 4]
```diff
@@ -14102,6 +14102,13 @@ TreeTransform<Derived>::TransformCXXTemporaryObjectExpr(
     if (TransformExprs(E->getArgs(), E->getNumArgs(), true, Args,
                        &ArgumentChanged))
       return ExprError();
+
+    if (E->isListInitialization() && !E->isStdInitListInitialization()) {
```
Comment (yronglin) [inline clang/lib/Sema/TreeTransform.h:14106]: If it is list initialization, but the constructor is called, and it is not the case of std::initializer_list, we should recreate an InitListExpr. https://godbolt.org/z/K4rzxosx3

File: clang/lib/Sema/SemaExpr.cpp (hunk 1 above)
Comment (erichkeane) [reply clang/lib/Sema/SemaExpr.cpp:5720]: I'm not sure I get the question?  The SFINAETrap IS suppressing at least some diagnostics.  

We can see if there are already errors pre-transformation via `containsError` (or if it is nullptr, depending on RecoveryExpr).  So if there was  previously detected error, one of those should catch it.

**Review by erichkeane — COMMENTED** [review]
Not the best one to be reviewing this, but the code generally looks good to me.

File: clang/lib/Sema/SemaExpr.cpp (hunk 1 above)
Comment (yronglin) [reply clang/lib/Sema/SemaExpr.cpp:5720]: Here we add `SFINAETrap Trap(*this)`. It was used to resolve the issue which looks like https://github.com/llvm/llvm-project/blob/main/clang/test/CXX/dcl.decl/dcl.init/p14-0x.cpp .
A shorter reproducer:
```
class Private {
  Private(int); // expected-note {{here}}
public:
  Private();
};

class S {
  Private p = 42; // expected-error {{private constructor}}

  S() {} // expected-error {{call to deleted constructor of 'NoDefault'}} \
            expected-error {{must explicitly initialize the member 'e1' which does not have a default constructor}}
  S(int) {}
};

// FIXME: test the other forms which use copy-initialization

```

`Field->getInClassInitializer()` contains an `bad access error`, then we try to rebuild this initializer, the diagnostic will emit 3 times. (The result of `Field->getInClassInitializer()->containsErrors()` 3 times are `false`).
1st time: parser member initializer.
2nd time: After rebuild, Line 5726 `      Res = ConvertMemberDefaultInitExpression(Field, Res.get(), Loc);`
3rd time: After rebuild, Line 5726 `      Res = ConvertMemberDefaultInitExpression(Field, Res.get(), Loc);`

```
./error.cpp:8:15: error: field of type 'Private' has private constructor
    8 |   Private p = 42; // expected-error {{private constructor}}
      |               ^
./error.cpp:2:3: note: implicitly declared private here
    2 |   Private(int); // expected-note {{here}}
      |   ^
./error.cpp:8:15: error: field of type 'Private' has private constructor
    8 |   Private p = 42; // expected-error {{private constructor}}
      |               ^
./error.cpp:2:3: note: implicitly declared private here
    2 |   Private(int); // expected-note {{here}}
      |   ^
./error.cpp:8:15: error: field of type 'Private' has private constructor
    8 |   Private p = 42; // expected-error {{private constructor}}
      |               ^
./error.cpp:2:3: note: implicitly declared private here
    2 |   Private(int); // expected-note {{here}}
      |   ^
3 errors generated.
```

**Review by erichkeane — APPROVED** [review]
I think this looks ok, I'd love it if @cor3ntin took another look, but I'm as happy as before.

File: clang/test/SemaCXX/cxx11-default-member-initializers.cpp (hunk 2 above)
Comment (ZequanWu) [reply clang/test/SemaCXX/cxx11-default-member-initializers.cpp:58]: Thanks!

**Review by cor3ntin — APPROVED** [review]

**Comment by yronglin** [comment]
Thanks for the review!

**Comment by bgra8** [comment]
heads-up @yronglin: we've bisected many compilation breakages to this patch.
We're working on a reproducer.

**Comment by bgra8** [comment]
Here's the reproducer:

```c++
struct a {
  virtual int *b();
};
template <class c> struct d : a {
  int *b() { new c; }
};
int *e(a *);
struct f {
  char *g;
};
struct h {};
struct i {
  i(const f &);
  i(h);
};
struct l {
  i j = i({.g = ""});
};
int *k = e(new d<l>);
struct n : l {};
int *m = e(new d<n>);
```

Compilation command:
```
$ clang -c repro.cc
```

This compiles fine with  `clang-18`, trunk `gcc` and previous `clang` revision.

@yronglin can you please have a look?

**Comment by alexfh** [comment]
> Here's the reproducer:
> 
> ```c++
> struct a {
>   virtual int *b();
> };
> template <class c> struct d : a {
>   int *b() { new c; }
> };
> int *e(a *);
> struct f {
>   char *g;
> };
> struct h {};
> struct i {
>   i(const f &);
>   i(h);
> };
> struct l {
>   i j = i({.g = ""});
> };
> int *k = e(new d<l>);
> struct n : l {};
> int *m = e(new d<n>);
> ```

A bit clearer repro: https://gcc.godbolt.org/z/r9EdcoW8G

The new clang behavior seems wrong to me.

**Comment by bgra8** [comment]
I reverted this and the dependent patch based on https://github.com/llvm/llvm-project/pull/92527#issuecomment-2149120420 to bring back the repo back to green.

**Comment by yronglin** [comment]
@bgra8 @alexfh Thanks report this issue, I've a new PR to fix it, and added your reproducer into the test.

**Pull request merged**